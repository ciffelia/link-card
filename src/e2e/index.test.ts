import { describe, it } from 'vitest';
import ky from 'ky';

describe.concurrent(
  '/',
  () => {
    if (process.env.LINK_CARD_TEST_URL === undefined) {
      throw new Error('LINK_CARD_TEST_URL is not set.');
    }
    const baseUrl = new URL(process.env.LINK_CARD_TEST_URL);

    const cases = [
      'http://example.com/',
      'https://ogp.me/',
      'https://www.w3.org/',
      'https://ogp.me/foobarbaz',
      'https://ogp.me/logo.png',
      'http://example.invalid/',
    ] as const satisfies readonly string[];

    for (const url of cases) {
      it(`should return a LinkCard object for ${url}`, async ({ expect }) => {
        const endpointUrl = new URL(baseUrl);
        endpointUrl.searchParams.set('url', url);
        expect(await ky(endpointUrl).json()).toMatchSnapshot();
      });
    }

    it('should return 400 for requests with no url parameter', async ({
      expect,
    }) => {
      const resp = await ky(baseUrl, { throwHttpErrors: false });
      expect(resp.status).toBe(400);
    });

    it('should return 400 for requests with an invalid url parameter', async ({
      expect,
    }) => {
      const url = new URL(baseUrl);
      url.searchParams.set('url', 'invalid');

      const resp = await ky(url, { throwHttpErrors: false });
      expect(resp.status).toBe(400);
    });
  },
  {
    timeout: 10000,
    retry: 3,
  },
);
