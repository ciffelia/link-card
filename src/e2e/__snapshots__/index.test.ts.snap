// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`/ > should return a LinkCard object for http://example.com/ 1`] = `
{
  "faviconUrl": "http://example.com/favicon.ico",
  "title": "Example Domain",
  "url": "http://example.com/",
}
`;

exports[`/ > should return a LinkCard object for http://example.invalid/ 1`] = `
{
  "url": "http://example.invalid/",
}
`;

exports[`/ > should return a LinkCard object for https://ogp.me/ 1`] = `
{
  "description": "The Open Graph protocol enables any web page to become a rich object in a social graph.",
  "faviconUrl": "https://ogp.me/favicon.ico",
  "ogImageUrl": "https://ogp.me/logo.png",
  "title": "Open Graph protocol",
  "url": "https://ogp.me/",
}
`;

exports[`/ > should return a LinkCard object for https://ogp.me/foobarbaz 1`] = `
{
  "faviconUrl": "https://ogp.me/favicon.ico",
  "url": "https://ogp.me/foobarbaz",
}
`;

exports[`/ > should return a LinkCard object for https://ogp.me/logo.png 1`] = `
{
  "faviconUrl": "https://ogp.me/favicon.ico",
  "url": "https://ogp.me/logo.png",
}
`;

exports[`/ > should return a LinkCard object for https://www.w3.org/ 1`] = `
{
  "description": "The World Wide Web Consortium (W3C) develops standards and guidelines to help everyone build a web based on the principles of accessibility, internationalization, privacy and security.",
  "faviconUrl": "https://www.w3.org/favicon.ico",
  "ogImageUrl": "https://www.w3.org/assets/website-2021/images/w3c-opengraph-image.png",
  "title": "W3C",
  "url": "https://www.w3.org/",
}
`;
